src/output-json-alert.c

```diff
@@ -230,9 +239,49 @@ static void AlertJsonSourceTarget(const Packet *p, const PacketAlert *pa,
     json_object_set_new(ajs, "target", tjs);
 }
 
+static int RuleMetadataLogKeyAsArray(AlertJsonOutputCtx *json_output_ctx, const char *key)
+{
+    KeysList *lkey = NULL;
+    TAILQ_FOREACH(lkey, &json_output_ctx->keys_list, next) {
+        if (!strcmp(lkey->key, key)) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static void AlertJsonMetadata(AlertJsonOutputCtx *json_output_ctx, const PacketAlert *pa, json_t *ajs)
+{
+    if (pa->s->metadata) {
+        DetectMetadata* kv = pa->s->metadata;
+        json_t *mjs = json_object();
+        while (kv) {
+            if ((json_output_ctx->flags & LOG_JSON_RULE_METADATA_AS_ARRAY) ||
+                RuleMetadataLogKeyAsArray(json_output_ctx, kv->key)) {
+                json_t *jkey = json_object_get(mjs, kv->key);
+                if (jkey == NULL) {
+                    jkey = json_array();
+                    if (jkey == NULL)
+                        return;
+                    json_array_append_new(jkey, json_string(kv->value));
+                    json_object_set_new(mjs, kv->key, jkey);
+                } else {
+                    json_array_append_new(jkey, json_string(kv->value));
+                }
+            } else {
+                json_object_set_new(mjs, kv->key, json_string(kv->value));
+            }
+            kv = kv->next;
+        }
+        json_object_set_new(ajs, "metadata", mjs);
```

Comment:
I don't think it's a good idea to have a object 'metadata' that can be either a map or an array. How would parsers deal with that?

Reply:
Seems a pure array approach makes much more sense to me still. It would work w/o config for all metadata fields, not just for the one the user had to configure.

Reply:
Metadata is always a map but some keys in the object can be depending on the key value a array or a single value.

```

Reply:
Ok it's not metadata but per field, the issue remains the same. Parsers can't assume they will know the type.

Whats wrong with a straight array? It's simpler and requires no config.

Reply:
I can see what @regit is trying to do here... Make it a scalar where it makes sense, and a list where that makes sense.. Based on the current status of metadata found in rules. Unfortunately I don't feel we can go ahead with such a thing without a formal specification/schema of what metadata should look like, and assign data types to the metadata types.

I also don't want to see it get hardcoded into the config, its hard to update and do the best thing over time.

Which unfortunately leaves arrays as the lowest common denominator.

I'm late to reviewing this PR, so still digesting it.

Reply:
With array you need to have array access in all your searches so this is quite an overload in term of processing and coding.